source/slang/slang-capability.cpp

```diff
@@ -170,121 +158,107 @@ bool isCapabilityDerivedFrom(CapabilityAtom atom, CapabilityAtom base)
+    for (auto i : out)
+        return (CapabilityAtom)i;
+    return CapabilityAtom::Invalid;
 }
 
-void CapabilitySet::addToTargetCapabilityWithTargetAndStageAtom(CapabilityName target, CapabilityName stage, const ArrayView<CapabilityName>& canonicalRepresentation)
+CapabilityAtom getStageAtomInSet(const CapabilityAtomSet& atomSet)
 {
-    // If no provided 'stage', set the capability as a target of all stages
-    if (stage == CapabilityName::Invalid)
-    {
-        auto info = _getInfo(CapabilityName::any_stage);
-        List<CapabilityName> newArr;
-        auto count = canonicalRepresentation.getCount();
-        newArr.setCount(count + 1);
-        memcpy(newArr.getBuffer(), canonicalRepresentation.getBuffer(), count * sizeof(CapabilityName));
-        m_targetSets[asAtom(target)].shaderStageSets.reserve(info.canonicalRepresentation.getCount());
-        for (auto i : info.canonicalRepresentation)
-        {
-            newArr[count] = i[0];
-            addToTargetCapabilityWithTargetAndStageAtom(target, i[0], newArr.getArrayView());
-        }
-        return;
-    }
-    
-    CapabilityAtomSet setToAdd = CapabilityAtomSet((UInt)CapabilityAtom::Count);
-    for(auto i : canonicalRepresentation)
-        setToAdd.add(asAtomUInt(i));
-
-    addToTargetCapabilityWithValidUIntSetAndTargetAndStage(target, stage, setToAdd);
+    auto stageSet = getAtomSetOfStages();
+    CapabilityAtomSet out;
+    CapabilityAtomSet::calcIntersection(out, stageSet, atomSet);
+    for(auto i : out)
+        return (CapabilityAtom)i;
+    return CapabilityAtom::Invalid;
 }
+#if defined(_MSC_VER)
+#pragma warning(pop)
+#endif
 
-// No targets atoms have been defined on yet, set stage to target any_target capability 
-void CapabilitySet::addToTargetCapabilityWithStageAtom(CapabilityName stage, const ArrayView<CapabilityName>& canonicalRepresentation)
+template<CapabilityName keyholeAtomToPermuteWith>
+void CapabilitySet::addPermutationsOfConjunctionForEachInContainer(CapabilityAtomSet& setToPermutate, const CapabilityAtomSet& elementsToPermutateWith, CapabilityAtom knownTargetAtom, CapabilityAtom knownStageAtom)
 {
-    
-    if (m_targetSets.getCount() == 0)
+    for(auto i : elementsToPermutateWith)
     {
-        const auto anyTargetInfo = _getInfo(CapabilityName::any_target);
-        CapabilityAtomSet setToAdd;
-        setToAdd.resize((UInt)CapabilityAtom::Count);
-        for (int i = 0; i < canonicalRepresentation.getCount(); i++)
-            setToAdd.add((UInt)canonicalRepresentation[i]);
-        CapabilityName targetAtom{};
-        for (const auto& targetAtomCanonicalRep : anyTargetInfo.canonicalRepresentation)
+        CapabilityName atom = (CapabilityName)i;
+        CapabilityAtomSet conjunctionPermutation = setToPermutate;
+        auto targetInfo = _getInfo(atom);
+        conjunctionPermutation.add(*targetInfo.canonicalRepresentation[0]);
+
+        if constexpr (keyholeAtomToPermuteWith == CapabilityName::target)
```

Comment:
Is it necessary to use `constexpr` here?
It wouldn't break any but it doesn't look like it will do any either.

Reply:
`if` statments are not ensured to resolve at compile time unless `constexpr` (even if the parameters are clearly known at compile time).

In practice this change will at the very least ensure the debug build resolves the `if` at compile time.

Reply:
Are you sure that the condition wouldn't be resolved at the compile time?
```
(keyholeAtomToPermuteWith == CapabilityName::target)
```
I will be very surprised if this wouldn't be resolved at a compile time.
We heavily relied on this type of compile time resolving feature to optimize the for-loop on PS3.

The keyword wouldn't hurt to have here.
But it is not a good practice to use keyword inappropriately.
It draws an unnecessary attention for the other engineers who will need to read the lines.

Reply:
> Are you sure that the condition wouldn't be resolved at the compile time?

That is not the reason `constexpr` exists in C++. Anything that is compile time evaluable can be evaluated by the compiler.
`constexpr` enforces behavior (or errors) to no longer rely on undefined compiler behavior.

I agree it likely will be evaluated at least in release. Debug builds likely will not. I can remove the `constexpr` since it is not required.

Reply:
turns out it is a warning to not use `constexpr` here

Reply:
I think you said it backward.
But anyway, if it prints a warning, let's keep it.

Reply:
> I think you said it backward. 

I don't see the problem in what I wrote?

Reply:
Sorry. I misread it.
Thanks for fixing the indentation at the next if-condition too.